cloud/linode/route_controller.go

```diff
@@ -0,0 +1,252 @@
+package linode
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/linode/linodego"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+
+	"github.com/linode/linode-cloud-controller-manager/cloud/linode/client"
+)
+
+type routeCache struct {
+	sync.RWMutex
+	routes     map[int][]linodego.VPCIP
+	lastUpdate time.Time
+	ttl        time.Duration
+}
+
+func (rc *routeCache) refreshRoutes(ctx context.Context, client client.Client) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if time.Since(rc.lastUpdate) < rc.ttl {
+		return nil
+	}
+
+	vpcNodes := map[int][]linodego.VPCIP{}
+	vpcID := vpcInfo.getID()
+	resp, err := client.ListVPCIPAddresses(ctx, vpcID, linodego.NewListOptions(0, ""))
+	if err != nil {
+		return err
+	}
+	for _, r := range resp {
+		vpcNodes[r.LinodeID] = append(vpcNodes[r.LinodeID], r)
+	}
+
+	rc.routes = vpcNodes
+	rc.lastUpdate = time.Now()
+	return nil
+}
+
+type routes struct {
+	vpcid      int
+	client     client.Client
+	instances  *instances
+	routeCache *routeCache
+}
+
+func newRoutes(client client.Client) (cloudprovider.Routes, error) {
+	timeout := 60
+	if raw, ok := os.LookupEnv("LINODE_ROUTES_CACHE_TTL"); ok {
+		if t, _ := strconv.Atoi(raw); t > 0 {
+			timeout = t
+		}
+	}
+	klog.V(3).Infof("TTL for routeCache set to %d", timeout)
+
+	vpcid := vpcInfo.getID()
```

Comment:
We can. I thought in future other controllers like node_controller or service_controller etc might also want to use vpcid and then we would be passing the same variable everywhere. Since its kind of global value shared by all and set once when the program inits, I kept it as a global var. Let me know if you still think passing it would be better and I can add the change.

Reply:
I am going to keep it as it is for now as in future, we might want to auto detect the VPC where CCM is running and in such cases, having a global variable and setting VPC id there would be useful for other controllers to consume once it is determined.